File: plugins/java/Freedom/src/main/java/com/github/kubinka/Freedom.java
package com.kubinka.freedom;

import android.content.Context;

import com.aliucord.annotations.AliucordPlugin;
import com.aliucord.entities.Plugin;
import com.aliucord.patcher.Hook;
import com.aliucord.patcher.PatcherAPI;

import java.util.HashMap;
import java.util.Map;

@AliucordPlugin
public class Freedom extends Plugin {

    private final Map<Character, Character> homoglyphs = new HashMap<Character, Character>() {{
        put('a', 'а'); put('e', 'е'); put('o', 'о'); put('p', 'р');
        put('c', 'с'); put('y', 'у'); put('x', 'х');
        put('A', 'А'); put('E', 'Е'); put('O', 'О'); put('P', 'Р');
        put('C', 'С'); put('X', 'Х');
    }};

    private String replaceHomoglyphs(String input) {
        StringBuilder sb = new StringBuilder();
        for (char ch : input.toCharArray()) {
            sb.append(homoglyphs.getOrDefault(ch, ch));
        }
        return sb.toString();
    }

    @Override
    public void start(Context ctx) throws Exception {
        Class<?> clazz = Class.forName("com.discord.models.domain.ModelMessageSendRequest");

        patcher.after(clazz, "getContent", new Class[0], (Hook) call -> {
            String original = (String) call.getResult();
            String modified = replaceHomoglyphs(original);
            call.setResult(modified);
            return null;
        });
    }

    @Override
    public void stop(Context ctx) {
        patcher.unpatchAll();
    }
}

File: plugins/kotlin/Freedom/src/main/kotlin/com/github/kubinka/Freedom.kt
package dev.kubinka0505.freedom

import android.content.Context
import com.discord.api.message.Message
import com.discord.models.domain.ModelMessageSendRequest
import de.robv.android.xposed.XC_MethodHook
import dev.alucord.annotations.AliucordPlugin
import dev.alucord.base.Plugin
import dev.alucord.patcher.after

@AliucordPlugin
class main : Plugin() {
    private val homoglyphs = mapOf(
        'a' to 'а', 'e' to 'е', 'o' to 'о', 'p' to 'р', 'c' to 'с', 'y' to 'у', 'x' to 'х',
        'A' to 'А', 'E' to 'Е', 'O' to 'О', 'P' to 'Р', 'C' to 'С', 'X' to 'Х'
    )

    private fun replaceHomoglyphs(input: String): String {
        return input.map { char -> homoglyphs[char] ?: char }.joinToString("")
    }

    override fun start(ctx: Context) {
        patcher.after<ModelMessageSendRequest>(
            "getContent"
        ) { call ->
            val original = call.result as String
            val modified = replaceHomoglyphs(original)
            call.setResult(modified)
        }
    }

    override fun stop(ctx: Context) {
        patcher.unpatchAll()
    }
}
